src/analysis/schwartz_space.lean

```diff
@@ -0,0 +1,383 @@
+  exact iterated_fderiv_neg_apply,
+end
+
+variables [normed_field 𝕜] [normed_space 𝕜 F] [smul_comm_class ℝ 𝕜 F]
+
+lemma decay_smul_aux (k n : ℕ) (f : 𝓢(E, F)) (c : 𝕜) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n (c • f) x∥ =
+  ∥c∥ * ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ :=
+by rw [mul_comm (∥c∥), mul_assoc, iterated_fderiv_const_smul_apply (f.smooth _), norm_smul]
+
+end aux
+
+section seminorm_aux
+
+/-- Helper definition for the seminorms of the Schwartz space. -/
+@[protected]
+def seminorm_aux (k n : ℕ) (f : 𝓢(E, F)) : ℝ :=
+Inf {c | 0 ≤ c ∧ ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c}
+
+lemma seminorm_aux_nonneg (k n : ℕ) (f : 𝓢(E, F)) : 0 ≤ f.seminorm_aux k n :=
+le_cInf (bounds_nonempty k n f) (λ _ ⟨hx, _⟩, hx)
+
+lemma le_seminorm_aux (k n : ℕ) (f : 𝓢(E, F)) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n ⇑f x∥ ≤ f.seminorm_aux k n :=
+le_cInf (bounds_nonempty k n f) (λ y ⟨_, h⟩, h x)
+
+/-- If one controls the norm of every `A x`, then one controls the norm of `A`. -/
+lemma seminorm_aux_le_bound (k n : ℕ) (f : 𝓢(E, F)) {M : ℝ} (hMp: 0 ≤ M)
+  (hM : ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ M) :
+  f.seminorm_aux k n ≤ M :=
+cInf_le (bounds_bdd_below k n f) ⟨hMp, hM⟩
+
+end seminorm_aux
+
+/-! ### Algebraic properties -/
+
+section smul
+
+variables [normed_field 𝕜] [normed_space 𝕜 F] [smul_comm_class ℝ 𝕜 F]
+  [normed_field 𝕜'] [normed_space 𝕜' F] [smul_comm_class ℝ 𝕜' F]
+
+instance : has_smul 𝕜 𝓢(E, F) :=
+⟨λ c f, { to_fun := c • f,
+  smooth' := (f.smooth _).const_smul c,
+  decay' := λ k n, begin
+    refine ⟨f.seminorm_aux k n * (∥c∥+1), λ x, _⟩,
+    have hc : 0 ≤ ∥c∥ := by positivity,
+    refine le_trans _ ((mul_le_mul_of_nonneg_right (f.le_seminorm_aux k n x) hc).trans _),
+    { apply eq.le,
+      rw [mul_comm _ (∥c∥), ← mul_assoc],
+      exact decay_smul_aux k n f c x },
+    { apply mul_le_mul_of_nonneg_left _ (f.seminorm_aux_nonneg k n),
+      linarith }
+  end}⟩
+
+@[simp] lemma smul_apply {f : 𝓢(E, F)} {c : 𝕜} {x : E} : (c • f) x = c • (f x) := rfl
+
+instance
+[has_smul 𝕜 𝕜'] [is_scalar_tower 𝕜 𝕜' F] : is_scalar_tower 𝕜 𝕜' 𝓢(E, F) :=
+⟨λ a b f, ext $ λ x, smul_assoc a b (f x)⟩
+
+instance [smul_comm_class 𝕜 𝕜' F] : smul_comm_class 𝕜 𝕜' 𝓢(E, F) :=
+⟨λ a b f, ext $ λ x, smul_comm a b (f x)⟩
+
+lemma seminorm_aux_smul_le (k n : ℕ) (c : 𝕜) (f : 𝓢(E, F)) :
+  (c • f).seminorm_aux k n ≤ ∥c∥ * f.seminorm_aux k n :=
+begin
+  refine (c • f).seminorm_aux_le_bound k n (mul_nonneg (norm_nonneg _) (seminorm_aux_nonneg _ _ _))
+    (λ x, (decay_smul_aux k n f c x).le.trans _),
+  rw mul_assoc,
+  exact mul_le_mul_of_nonneg_left (f.le_seminorm_aux k n x) (norm_nonneg _),
+end
+
+instance has_smul_nat : has_smul ℕ 𝓢(E, F) :=
+⟨λ c f, { to_fun := c • f,
+  smooth' := (f.smooth _).const_smul c,
+  decay' := begin
+    have : c • (f : E → F) = (c : ℝ) • f,
+    { ext x, simp only [pi.smul_apply, ← nsmul_eq_smul_cast] },
+    simp only [this],
+    exact ((c : ℝ) • f).decay',
+  end}⟩
+
+instance has_smul_int : has_smul ℤ 𝓢(E, F) :=
```

Comment:
I'd expect these to be named `has_nsmul` and `has_zsmul`, but maybe these are just historical artefacts so if you don't like it that's not a problem